File: example-projects/csharp-basic/src/Services/OrderService.cs
using CSharpBasic.Models;
using CSharpBasic.Utils;

namespace CSharpBasic.Services;

public class OrderService : IOrderService
{
    private readonly Dictionary<int, Order> _store = [];
    private int _nextId = 1;

    public Order PlaceOrder(string customerName, IEnumerable<(Product product, int quantity)> items)
    {
        var lines = items.Select(i => new OrderLine { Product = i.product, Quantity = i.quantity }).ToList();
        var order = new Order { Id = _nextId++, CustomerName = customerName, Lines = lines };
        _store[order.Id] = order;
        return order;
    }

    public Order? GetOrder(int id) => _store.GetValueOrDefault(id);

    public IEnumerable<Order> GetAllOrders() => _store.Values;

    public bool CancelOrder(int id) => _store.Remove(id);

    public decimal GetDiscountedTotal(int id, decimal discountPercent)
    {
        var order = GetOrder(id) ?? throw new KeyNotFoundException($"Order {id} not found.");
        return PriceCalculator.ApplyDiscount(order.Total, discountPercent);
    }
}

File: example-projects/rust-basic/src/services/order_service.rs
use std::collections::HashMap;
use crate::models::{Order, Product};
use crate::utils::price_calculator;

pub trait OrderServiceTrait {
    fn place_order(&mut self, customer_name: &str, items: Vec<(Product, u32)>) -> &Order;
    fn get_order(&self, id: u32) -> Option<&Order>;
    fn all_orders(&self) -> Vec<&Order>;
    fn cancel_order(&mut self, id: u32) -> bool;
}

pub struct OrderService {
    store: HashMap<u32, Order>,
    next_id: u32,
}

impl OrderService {
    pub fn new() -> Self {
        Self { store: HashMap::new(), next_id: 1 }
    }

    pub fn discounted_total(&self, id: u32, discount_percent: f64) -> Option<f64> {
        self.get_order(id).map(|o| price_calculator::apply_discount(o.total(), discount_percent))
    }
}

impl OrderServiceTrait for OrderService {
    fn place_order(&mut self, customer_name: &str, items: Vec<(Product, u32)>) -> &Order {
        let id = self.next_id;
        self.next_id += 1;
        let mut order = Order::new(id, customer_name);
        for (product, qty) in items {
            order.add_line(product, qty);
        }
        self.store.insert(id, order);
        self.store.get(&id).unwrap()
    }

    fn get_order(&self, id: u32) -> Option<&Order> {
        self.store.get(&id)
    }

    fn all_orders(&self) -> Vec<&Order> {
        self.store.values().collect()
    }

    fn cancel_order(&mut self, id: u32) -> bool {
        self.store.remove(&id).is_some()
    }
}
